natchez-doobie/src/test/scala/com/ovoenergy/effect/natchez/TracedTransactorTest.scala

```diff
@@ -0,0 +1,59 @@
+package com.ovoenergy.effect.natchez
+
+import cats.effect.concurrent.Ref
+import cats.effect.{Blocker, ContextShift, IO, Resource}
+import cats.syntax.functor._
+import com.ovoenergy.effect.natchez.TracedTransactor.Traced
+import doobie.h2.H2Transactor.newH2Transactor
+import doobie.implicits._
+import doobie.util.transactor.Transactor
+import natchez.{Kernel, Span, TraceValue}
+import org.scalatest.{Matchers, WordSpec}
+import cats.syntax.flatMap._
+
+import scala.concurrent.ExecutionContext.global
+
+class TracedTransactorTest extends WordSpec with Matchers {
+  implicit val cs: ContextShift[IO] = IO.contextShift(global)
+
+  case class SpanData(
+    name: String,
+    tags: Map[String, TraceValue]
+  )
+
+  def run[A](a: Traced[IO, A]): IO[List[SpanData]] =
+    Ref.of[IO, List[SpanData]](List(SpanData("root", Map.empty))).flatMap { sps =>
+      lazy val spanMock: Span[IO] = new Span[IO] {
+        def span(name: String): Resource[IO, Span[IO]] =
+          Resource.liftF(sps.update(_ :+ SpanData(name, Map.empty)).as(spanMock))
+        def kernel: IO[Kernel] =
+          IO.pure(Kernel(Map.empty))
+        def put(fields: (String, TraceValue)*): IO[Unit] =
+          sps.update(s => s.dropRight(1) :+ s.last.copy(tags = s.last.tags ++ fields.toMap))
+      }
+      a.run(spanMock).attempt.flatMap(_ => sps.get)
+    }
+
+  val db: Resource[IO, Transactor[Traced[IO, *]]] =
+    for {
+      block <- Blocker[IO]
+      xa <- newH2Transactor[IO]("jdbc:h2:mem:test","foo", "bar", global, block)
+    } yield TracedTransactor("test", block, xa)
+
+  "TracedTransactor" should {
+
+    "Trace queries" in {
```

Comment:
if the query had stuff that should not be logged (eg, user name or something) that would be a problem or the prepared statement would hide all these with `?`?

Reply:
As long as they're substituted in with string interpolation I think we'll be fine, it is quite fortunate